test/common/alerts_firing.go

```diff
@@ -0,0 +1,135 @@
+package common
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	prometheusv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+)
+
+const deadMansSwitch = "DeadMansSwitch"
+
+// alertsTestMetadata contains metadata about the alert
+type alertTestMetadata struct {
+	alertName string
+	podName   string
+	namespace string
+}
+
+// alertsFiringError is a custom alerts error
+type alertsFiringError struct {
+	alertsPending        []alertTestMetadata
+	alertsFiring         []alertTestMetadata
+	deadMansSwitchFiring bool
+}
+
+// Error implements the error interface and returns a readable output message
+func (e *alertsFiringError) Error() string {
+	var str strings.Builder
+	if e.deadMansSwitchFiring {
+		str.WriteString("\nThe following alerts were not fired, but were expected to be firing:")
+		str.WriteString(fmt.Sprintf("\n\talert: %s", deadMansSwitch))
+	}
+	if len(e.alertsPending) != 0 {
+		str.WriteString("\nThe following alerts were pending:")
+		for _, alert := range e.alertsPending {
+			str.WriteString(fmt.Sprintf("\n\talert: %s, for pod: %s, in namespace: %s", alert.alertName, alert.podName, alert.namespace))
+		}
+	}
+	if len(e.alertsFiring) != 0 {
+		str.WriteString("\nThe following alerts were fired:")
+		for _, alert := range e.alertsFiring {
+			str.WriteString(fmt.Sprintf("\n\talert: %s, for pod: %s, in namespace: %s", alert.alertName, alert.podName, alert.namespace))
+		}
+	}
+	return str.String()
+}
+
+// isNotEmpty checks whether or not the error contains firing or pending alerts
+func (e *alertsFiringError) isNotEmpty() bool {
+	return !e.deadMansSwitchFiring || len(e.alertsPending) != 0 || len(e.alertsPending) != 0
+}
+
+func TestIntegreatlyAlertsFiring(t *testing.T, ctx *TestingContext) {
+	var lastError error
+
+	// retry the tests every minute for up to 15 minutes
+	monitoringTimeout := 15 * time.Minute
+	monitoringRetryInterval := 1 * time.Minute
+	err := wait.Poll(monitoringRetryInterval, monitoringTimeout, func() (done bool, err error) {
+		if newErr := getFiringOrPendingAlerts(ctx); newErr != nil {
+			lastError = newErr
+			if _, ok := newErr.(*alertsFiringError); ok {
+				return false, nil
+			}
+			return false, err
```

Comment:
If we return `newErr` there, the poll function will exit immediately and won't continue retrying. So I just save `newErr` to the `lastErr` variable and return that if there's an error at the very end

Reply:
From my understand `err` is just `nil` at this point because I can't see where it's initialized, also from my understanding if `newError` is not of type `alertsFiringError` it means that the `getFiringOrPendingAlerts` function is failing for some other reason (line 86, 93, ..) so in that case, is fine to fail immediately because the result would not change if repeating the operation again, but maybe I'm missing something here

Reply:
Ahhhh facepalm! Thought you were talking about line 68, you're completely right. Great catch, thank you!

Reply:
`err` here is the one that is declared in the "return" part of the anonymous function signature. It's `nil` because that's the zero value for the error type, and it's not set to anything else.

To @b1zzu's point, is there value in retrying for 15 minutes if e.g. it fails to exec into the prometheus pod? I think there might be in some circumstances at least (it got restarted or rescheduled and will be back momentarily, or there was a blip in the k8s api server or something). If we're retrying on this though, it would be good to know, so maybe the error should be logged to the console here in a message saying that it will retry.

Also, for our future selves looking making changes to this, I think it would be slightly clearer if it was just `return false, nil` -- IMHO the named return values are only really useful if using naked returns (and I'd hesitate to use naked returns!).

Reply:
@grdryn Agreed, that was actually just an oversight when I renamed `err` to `newErr`, I accidentally missed that line..

Reply:
Yes I agree if we really want to retry for all errors than using `return false, nil` is more explicit, but at that point, there is no difference between line 68 and 70

Reply:
@b1zzu yeah that's true too I guess! :) @dimitraz if that's the intention, then that loop can maybe disappear?

Reply:
I would personally prefer not to retry for all errors, just alerting ones

Reply:
Sorry for the confusion guys

Reply:
I don't have a strong opinion either way, so sounds good :+1: